elasticsearch_django/models.py

```diff
@@ -271,8 +277,14 @@ def update_search_index(self, action, index='_all', force=False):
         Returns the HTTP response.
 
         """
-        assert action in ('index', 'update', 'delete'), ("Action must be 'index', 'update' or 'delete'.")  # noqa
-        assert self.pk, "Object must have a primary key before being indexed."
+        if action not in ('index', 'update', 'delete'):
+            raise ValueError("Action must be 'index', 'update' or 'delete'.")
+        if not self.pk:
+            raise ValueError("Object must have a primary key before being indexed.")
+        if action == 'update' and not update_fields:
+            raise ValueError(
+                "update_fields must be set to a non-empty list of strings when action is 'update'"
+            )
```

Comment:
Open to discussion but I feel like this is the wrong default, as it goes against how Django operates in terms of its own `update_fields` argument; I'm worried it could introduce confusion as to how it operates.

With Django, you can `save` without specifying an `update_fields`, and in that case it will just save the entire object. Rather than requiring `update_fields` to be set, we could follow Django and assume that if you're calling for an `update` action without setting `update_fields`, that you want the entire object updated.

Thoughts?

Reply:
I've now changed the error message to include a suggestion to use 'index' instead of 'update' if the user wants to perform a full update.

Reply:
Yup - looking at this again I think you're right @djm. That said, I think we should log any instances of this (`'update' and not update_fields`) - as we are asserting (in the english sense, not python) that an update is a partial - which is how the underlying `Elasticsearch.update` works.

```python
if action == 'update' and not update_fields:
    logger.warning("Switching action from 'update' to 'index' as `update_fields` is not specified.")
    action == 'index'
```